File: leetcode/0331-verify-preorder-serialization-of-a-binary-tree/solution.py
class Solution:
    def isValidSerialization(self, preorder: str) -> bool:
        nodes = preorder.split(',')
        slots = 1  # number of open "slots" available for the next node

        for node in nodes:
            if slots <= 0:
                return False
            slots -= 1  # this node fills one open slot
            if node != '#':
                slots += 2  # a real node opens two new slots for its children

        return slots == 0

File: leetcode/0331-verify-preorder-serialization-of-a-binary-tree/solution.cpp
#include <string>
#include <sstream>
using namespace std;

class Solution {
public:
    bool isValidSerialization(string preorder) {
        int slots = 1;
        stringstream ss(preorder);
        string node;

        while (getline(ss, node, ',')) {
            if (slots <= 0) {
                return false;
            }
            slots--;
            if (node != "#") {
                slots += 2;
            }
        }

        return slots == 0;
    }
};
